tracer/src/Datadog.Trace/SpanContext.cs

```diff
@@ -314,6 +318,44 @@ IEnumerator IEnumerable.GetEnumerator()
             }
         }
 
+        /// <summary>
+        /// Sets a DataStreams checkpoint
+        /// </summary>
+        /// <param name="manager">The <see cref="DataStreamsManager"/> to use</param>
+        /// <param name="edgeTags">The edge tags for this checkpoint. NOTE: These MUST be sorted alphabetically</param>
+        internal void SetCheckpoint(DataStreamsManager manager, string[] edgeTags)
+        {
+            PathwayContext = manager.SetCheckpoint(PathwayContext, edgeTags);
+        }
+
+        /// <summary>
+        /// Merges two DataStreams <see cref="PathwayContext"/>
+        /// Should be called when a pathway context is extracted from an incoming span
+        /// Used to merge contexts in a "fan in" scenario.
+        /// </summary>
+        internal void MergePathwayContext(PathwayContext? pathwayContext)
+        {
+            if (pathwayContext is null)
+            {
+                return;
+            }
+
+            if (PathwayContext is null)
+            {
+                PathwayContext = pathwayContext;
+                return;
+            }
+
+            // This is purposely not thread safe
+            // The code randomly chooses between the two PathwayContexts.
+            // If there is a race, then that's okay
+            // Randomly select between keeping the current context (0) or replacing (1)
+            if (ThreadSafeRandom.Next(2) == 1)
```

Comment:
_Mostly_ because that's what the other implementations do 😅 But I think of it like this: basically, you have 2 "paths" to this point, e.g. you're consuming multiple messages and "aggregating" them into another message. We can only track a single path at a time, so this "merge" step ignores one of the paths. Essentially we sample each path 50%. If we tried to track them both, we would have to propagate _both_ to downstream services, which in turn would have to calculate everything for both paths etc leading to exponential growth in header size and calcs etc.

Reply:
That makes sense thanks!